File: lab2/Money.cpp
#include "Money.h"
#include <stdexcept>

Money::Money(int _firstInt, int _secondInt) : Pair(_firstInt, _secondInt)
{
	if (_secondInt > MAX_SECOND)
	{
		_firstInt += _secondInt / MAX_SECOND;
		_secondInt = _secondInt % MAX_SECOND;
	}

	if (_firstInt < 0 || _secondInt < 0)
	{
		throw std::logic_error("Less than zero\n");
	}
}

Money* Money::Plus(const Pair& source)
{
	int firstInt = GetFirstInt();
	int secondInt = source.GetSecondInt();
	_firstInt += firstInt + (_secondInt + secondInt) / MAX_SECOND;
	_secondInt = (_secondInt + secondInt) % MAX_SECOND;
	return this;
}

Money* Money::Minus(const Pair& source)
{
	int firstInt = source.GetFirstInt();
	int secondInt = source.GetSecondInt();

	_firstInt = (_firstInt > firstInt) ? _firstInt - firstInt : 0;

	if (_secondInt >= secondInt)
	{
		_secondInt -= secondInt;
	}
	else if (_firstInt)
	{
		_firstInt--;
		_secondInt = _secondInt - secondInt + MAX_SECOND;
	}
	else
	{
		_secondInt = 0;
	}

	return this;
}

string Money::GetPair() const
{

	string result = std::to_string(_firstInt) + "." + std::to_string(_secondInt) + " rub";
	return result;
}

File: lab5-6/lab6/lab6/Money.cs
using System;

namespace lab6
{
    class Money : Pair
    {

        public Money(int first, int second) : base(first, second)
        {
            if (_second > MAX_SECOND)
            {
                _first += _second / MAX_SECOND;
                _second = _second % MAX_SECOND;
            }

            if (_first < 0 || _second < 0)
            {
                throw new ArgumentOutOfRangeException("Less than zero\n");
            }
        }

        public override string GetPair()
        {
            string result = _first + "." + _second + " rub";
            return result;
        }

        public override Pair Minus(Pair source)
        {
            int firstInt = source.GetFirst;
            int secondInt = source.GetSecond;

            int resFirst = _first;
            int resSecond = _second;

            resFirst = (_first > firstInt) ? _first - firstInt : 0;

            if (_second >= secondInt)
            {
                resSecond = _second - secondInt;
            }
            else if (_first == 1)
            {
                resFirst = _first - 1;
                resSecond = _second - secondInt + MAX_SECOND;
            }
            else
            {
                resSecond = 0;
            }

            return new Money(resFirst, resSecond);
        }

        public override Pair Plus(Pair source)
        {
            int firstInt = source.GetFirst;
            int secondInt = source.GetSecond;

            int resFirst = _first;
            int resSecond = _second;

            resFirst += firstInt + (_second + secondInt) / MAX_SECOND;
            resSecond = (_second + secondInt) % MAX_SECOND;

            return new Money(resFirst, resSecond);
        }

        private readonly int MAX_SECOND = 100;
    }
}
